File: part_8-deep-learning/artificial_neural_network/ann.py
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('../../data_files/Churn_Modelling.csv')
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values

# ------ Part-1: Data preprocessing ----------

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categorical_features=[1])
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:]

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# ------- Part-2: Build the ANN --------

# import keras library and packages
import keras
from keras.models import Sequential
from keras.layers import Dense

# Initializing the ANN
classifier = Sequential()

# Adding the input layer and the first hidden layer
layer_info = Dense(activation='relu', input_dim=11, kernel_initializer='uniform', units=6)
classifier.add(layer_info)

# Adding second hidden layer
layer_info = Dense(activation='relu', kernel_initializer='uniform', units=6)
classifier.add(layer_info)

# Adding output layer
layer_info = Dense(activation='sigmoid', kernel_initializer='uniform', units=1)
classifier.add(layer_info)

# Compiling the ANN
classifier.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Fitting the ANN to the training set
classifier.fit(X_train, y_train, batch_size=10, nb_epoch=100)

# Predicting the Test set results
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)

# Making the confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

File: part_8-deep-learning/artificial_neural_network/ann.R
# Importing the dataset
# ---------------------
dataset = read.csv('../../data_files/Churn_Modelling.csv')
dataset = dataset[4:14]

# Encoding the categorical variables as factor
dataset$Geography = as.numeric(factor(dataset$Geography,
                                      levels=c('France', 'Spain', 'Germany'),
                                      labels=c(1, 2, 3)))
dataset$Gender = as.numeric(factor(dataset$Gender,
                                   levels=c('Female', 'Male'),
                                   labels=c(1, 2)))

# Splitting the dataset into the Training set and Test set
# --------------------------------------------------------
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Exited, SplitRatio = 0.8)
training_set = subset(dataset, split==TRUE)
test_set = subset(dataset, split==FALSE)

# Feature scaling
# ---------------
training_set[-11] = scale(training_set[-11])
test_set[-11] = scale(test_set[-11])

# Fitting ANN Classifier to training set
#---------------------------------------------
library(h2o)
h2o.init(nthreads=-1)
classifier = h2o.deeplearning(y='Exited',
                              training_frame=as.h2o(training_set),
                              activation='Rectifier',
                              hidden=c(6, 6),
                              epochs=100,
                              train_samples_per_iteration=-2)

# Predicting the Test set results
#--------------------------------
prob_pred = h2o.predict(classifier, newdata=as.h2o(test_set[-11]))
y_pred = (prob_pred > 0.5)
y_pred = as.vector(y_pred)

# Making the Confusion Matrix
cm = table(test_set[, 11], y_pred)

# Shutting down H2O instance
h2o.shutdown()
